### ISSUE TITLE
EMR 3.3.1 with spark 1.1.1e error in cluster mode

### ISSUE THREAD

**Issue by jkleckner** [awslabs/emr-bootstrap-actions#47]
I have been successful with YARN client mode using a 3.3.1 AMI and spark 1.1.1e script.

But when I run it in cluster mode I get the error message below.

There are old fixed errors that sound suspiciously similar as in:
- https://issues.apache.org/jira/browse/MAPREDUCE-3931
  - This fix found a long used as an atomic item and added synchronize to fix it.

I tried the removal of .sparkStaging as suggested here but that made no difference:
- http://apache-spark-user-list.1001560.n3.nabble.com/jar-changed-on-src-filesystem-td10011.html

It reliably has this error every time.

The command line looks like this:

```bash
/home/hadoop/spark/bin/spark-submit \
  --verbose \
  --deploy-mode cluster \
  --master yarn-cluster \
  --driver-memory 4G \
  --executor-memory 5G \
  --num-executors  5 \
  --jars /home/hadoop/test/test-assembly-0.1.0-SNAPSHOT.jar \
  --class test... \
  /home/hadoop/test/test-assembly-0.1.0-SNAPSHOT.jar \
  args
```

The variation that just uses client mode works fine as in:

```bash
/home/hadoop/spark/bin/spark-submit \
  --verbose \
  --deploy-mode client \
  --master yarn-client \
  --driver-memory 4G \
  --executor-memory 5G \
  --num-executors  5 \
  --jars /home/hadoop/test/test-assembly-0.1.0-SNAPSHOT.jar \
  --class test... \
  /home/hadoop/test/test-assembly-0.1.0-SNAPSHOT.jar \
  args
```

Error message:

```
2015-01-16 00:53:49,401 INFO  [main] yarn.Client (Logging.scala:logInfo(59)) - Application report from ResourceManager:
  application identifier: application_1421213387393_0069
  appId: 69
  clientToAMToken: null
  appDiagnostics: Application application_1421213387393_0069 failed 2 times due to AM Container for appattempt_1421213387393_0069_000002 exited with  exitCode: -1000 due to: Resource hdfs://xxxxxxxxxxxx:9000/user/hadoop/.sparkStaging/application_1421213387393_0069/xxxxx-assembly-0.1.0-SNAPSHOT.jar changed on src filesystem (expected 1421369607450, was 1421369608135
g this attempt.. Failing the application.
  appMasterHost: N/A
  appQueue: default
  appMasterRpcPort: -1
  appStartTime: 1421369608345
  yarnAppState: FAILED
  distributedFinalState: FAILED
  appTrackingUrl: xxxxxxxxxxxxxxx.ec2.internal:9026/cluster/app/application_1421213387393_0069
  appUser: hadoop
```

This is using the 3.3.1 AMI and spark 1.1.1e script.

**Comment by christopherbozeman** [comment]
Are you building your application against a copy of the spark-assembly-*.jar as found on the cluster in /home/hadoop/spark/lib/ and referencing it as a external jar?

**Comment by jkleckner** [comment]
Here is a redacted version of how the cluster is built.
```bash
aws emr create-cluster \
  --region us-east-1 \
  --name ${stackName} \
  --instance-groups '[
    { "Name": "spark_master", "InstanceCount": 1, "InstanceGroupType": "MASTER", '"${useSpot[*]}"' "InstanceType": "m3.'${stackSize}'" },
    { "Name": "spark_core",   "InstanceCount": 1, "InstanceGroupType": "CORE",   '"${useSpot[*]}"' "InstanceType": "m3.'${stackSize}'" },
    { "Name": "spark_slave",  "InstanceCount": 1, "InstanceGroupType": "TASK",   '"${useSpot[*]}"' "InstanceType": "m3.'${stackSize}'" }
  ]' \
  --no-auto-terminate \
  --log-uri s3://xxxxxxxxxxxxxxxxxxxxx \
  --ec2-attributes KeyName=xxxxxxxxxxxx,InstanceProfile=xxxxxxxxx \
  --no-termination-protected \
  --visible-to-all-users \
  --enable-debugging \
  --tags \
    Name=${stackName} \
  --applications Name=Ganglia \
  --bootstrap-actions \
    Name=InstallSpark,Args="-v1.1.1.e",Path=s3://support.elasticmapreduce/spark/install-spark \
    Name="Configure HBase for Ganglia",Path=s3://us-east-1.elasticmapreduce/bootstrap-actions/configure-hbase-for-ganglia \
  --steps \
    Name=SparkHistoryServer,Jar=s3://elasticmapreduce/libs/script-runner/script-runner.jar,Args=s3://support.elasticmapreduce/spark/start-history-server \
  --ami-version 3.3.1
```

**Comment by christopherbozeman** [comment]
I see the most likely cause.  The latest Spark version by default for AMI 3.3.1 is currently 1.2.0.    The arguments requested on the CLI are being ignored.

This `Args="-v1.1.1.e"` should be written as `Args=-v,1.1.1.e` for the AWS CLI.

Please give that a try.

**Comment by jkleckner** [comment]
Tried it and got the same error.

Separately, if I want -x and -g would the Args be `Args=-v,1.1.1.e,-x,-g` ?

**Comment by christopherbozeman** [comment]
That is correct about the additional args. Please provide a cluster id (j-####) and I can dig further.

**Comment by jkleckner** [comment]
j-3N0GFZYW1FELE

**Comment by christopherbozeman** [comment]
Thanks....I verified you are getting 1.1.1.e as expected and I was able to reproduce the error:

```
2015-01-16 23:57:13,366 INFO  [main] yarn.Client (Logging.scala:logInfo(59)) - Application report from ResourceManager:
         application identifier: application_1421452393704_0002
         appId: 2
         clientToAMToken: null
         appDiagnostics: Application application_1421452393704_0002 failed 2 times due to AM Container for appattempt_1421452393704_0002_000002 exited with  exitCode: -1000 due to: Resource hdfs://10.31.95.200:9000/user/hadoop/.sparkStaging/application_1421452393704_0002/spark-wordcount-1.1.1.jar changed on src filesystem (expected 1421452628103, was 1421452628149
.Failing this attempt.. Failing the application.
``

This is error is due to including the same application jar as an additional jar with `--jars`.   When in cluster mode Spark will already push the main application jar to the driver without having to explicitly declare it.  The reason this works in yarn-client mode is that Spark does not push it up since it only needs to run locally.

Thus give this a try:

```
/home/hadoop/spark/bin/spark-submit \
  --verbose \
  --deploy-mode cluster \
  --master yarn-cluster \
  --driver-memory 4G \
  --executor-memory 5G \
  --num-executors  5 \
  --class test... \
  /home/hadoop/test/test-assembly-0.1.0-SNAPSHOT.jar \
  args
```

**Comment by jkleckner** [comment]
I confirm that error becomes fixed.

I now see that the problem is that I need to propagate environment variables to the executors.

I guess ```spark.yarn.appMasterEnv.<envname>``` is the best way?

**Comment by christopherbozeman** [comment]
That property would only populate the driver environment which would be the application master. 

See https://spark.apache.org/docs/1.1.1/configuration.html

Use the spark.executorEnv.<envname> to pass to the executors.

**Comment by jkleckner** [comment]
Doh, of course executorEnv...

FYI, I already had a bootstrap step to put env values into ```/home/hadoop/spark/conf/spark-env.sh``` and thought it would be used for all instances.  I confirmed that the env settings were in that file for each type of executor.  And bootstrap should be in effect before starting those services, true?

So I tried adding a step to set spark.executorEnv.env for all the vars.  I confirmed that it is set on master after ```running```.  I don't see it propagated to the executor nodes though.  And it still failed with log message indicating that the env vars are not set.  Further ideas on where to look?

BTW, I'm very grateful for your help on this.  It has been invaluable.
